Fix exception in home_connect.light

antlarr opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

The line above the fixed condition might set brightness to an empty
dict if self.device.appliance.status doesn't have a brightness key.
The condition incorrectly checks exclusively for "brightness is None" so it's better to
check for "not brightness" which contains the previous check but also handles
correctly the empty dict case.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #86222

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

Note: the change is so small and contained that I haven't run tests locally since I guess tests don't cover that and they'll run before the PR is merged anyway.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

(I added "LGTM" comments in the two PRs I reviewed)

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

antlarr committed:
Fix exception in home_connect.light

The line above the fixed condition might set brightness to an empty
dict if self.device.appliance.status doesn't have a brightness key.
This result in the following exception being raised and shown in
the logs:

2023-01-19 06:32:03.146 ERROR (MainThread) [homeassistant.components.light] home_connect: Error on device update!
Traceback (most recent call last):
  File "/opt/homeassistant/lib/python3.9/site-packages/homeassistant/helpers/entity_platform.py", line 503, in _async_add_entity
    await entity.async_device_update(warning=False)
  File "/opt/homeassistant/lib/python3.9/site-packages/homeassistant/helpers/entity.py", line 729, in async_device_update
    await task
  File "/opt/homeassistant/lib/python3.9/site-packages/homeassistant/components/home_connect/light.py", line 200, in async_update
    self._brightness = ceil((brightness.get(ATTR_VALUE) - 10) * 255 / 90)
TypeError: unsupported operand type(s) for -: 'NoneType' and 'int'

thecode commented on homeassistant/components/home_connect/light.py:
```diff
@@ -194,7 +194,7 @@ async def async_update(self) -> None:
         else:
             brightness = self.device.appliance.status.get(self._brightness_key, {})
-            if brightness is None:
+            if not brightness:
```
Is 0 a valid value for brightness? if it is, it would be better to change:
`brightness = self.device.appliance.status.get(self._brightness_key)` 
and keep the original condition

antlarr replied:
I guess 0 is not a valid value since if the value from `self.device.appliance.status.get(self._brightness_key)` is returned correctly it's used later as `brightness.get(ATTR_VALUE)` which would fail if brightness was an integer. Isn't it?

antlarr replied:
I think that the brightness variable is not what you think. There are two cases, if `self.device.appliance.status.get(self._brightness_key, {})` returns a valid value or if the `self.device.appliance.status` dict doesn't have a `self._brightness_key` key and it would return the default value (`{}`).

- If the returned value is valid, then that's handled in line 200 (just a few lines below) and  `brightness.get(ATTR_VALUE)` would be executed. This means that in this case, brightness is a dict which would have an ATTR_VALUE key.

- If the returned value is not valid, then an empty dict would be assigned to brightness, which is the reason the check for `if brightness is None` was wrong and I changed it to `if not brightness`.

emontnemery commented on homeassistant/components/home_connect/light.py:
```diff
@@ -194,7 +194,7 @@ async def async_update(self) -> None:
         else:
             brightness = self.device.appliance.status.get(self._brightness_key, {})
-            if brightness is None:
+            if not brightness:
```
```suggestion
            if not brightness or ATTR_VALUE not in brightness:
```

emontnemery commented on homeassistant/components/home_connect/light.py:
```diff
@@ -194,7 +194,7 @@ async def async_update(self) -> None:
+            if not brightness:
                 self._brightness = None
             else:
                 self._brightness = ceil((brightness.get(ATTR_VALUE) - 10) * 255 / 90)
```
```suggestion
                self._brightness = ceil((brightness[ATTR_VALUE] - 10) * 255 / 90)
```

antlarr replied:
I think this changes the coding style of the file (which always uses get instead of the [] operator) so I guess the author/maintainer of the integration wouldn't like it.

antlarr replied:
Just a note for completion: I would also prefer the [] operator but I'm just trying to keep the patch simple and I'm used to not changing the coding style of other developers when submitting PRs.

frenck replied:
This has nothing to do with that. The line above you changed already ensures it is in the dict, so it can be referenced directly (without making an unneeded get() method call).

@antlarr Could you apply the suggestion by emontnemery above?

../Frenck

antlarr replied:
There are many cases of `.get(ATTR_VALUE)` in that file (and even more in other files in that component). Do you want me to change all of them to use `[ATTR_VALUE]` or only that one?

emontnemery commented on homeassistant/components/home_connect/light.py:
```diff
@@ -194,7 +194,7 @@ async def async_update(self) -> None:
 
         else:
             brightness = self.device.appliance.status.get(self._brightness_key, {})
```
```suggestion
            brightness = self.device.appliance.status.get(self._brightness_key)
```

emontnemery reviewed (commented):
I'd suggest to clean this up a bit further since the line involving `brightness.get` below will also throw if the wanted key is not there.

antlarr committed:
Check for ATTR_VALUE before using it

Update homeassistant/components/home_connect/light.py as suggested by @emontnemery in #86225

Co-authored-by: Erik Montnemery <erik@montnemery.com>

antlarr committed:
Don't use a default value now that None is handled correctly

Co-authored-by: Erik Montnemery <erik@montnemery.com>

ruudk commented:
Would it be possible to merge this PR? It's making it impossible for me to upgrade to the latest version.
